yarn-project/noir-compiler/src/compile.ts

```diff
@@ -0,0 +1,105 @@
+ */
+export interface Dependency {
+  /**
+   * Path to the dependency.
+   */
+  path?: string;
+  /**
+   * Git repository of the dependency.
+   */
+  git?: string;
+}
+
+/**
+ * A class that compiles noir contracts and outputs the Aztec ABI.
+ */
+export class ContractCompiler {
+  constructor(private projectPath: string) {}
+
+  /**
+   * Compiles the contracts in projectPath and returns the Aztec ABI.
+   * @returns Aztec ABI of the compiled contracts.
+   */
+  public async compile(): Promise<ContractAbi[]> {
+    const noirContracts = await this.compileNoir();
+    const abis: ContractAbi[] = noirContracts.map(this.convertToAztecABI);
+    return abis;
+  }
+
+  /**
+   * Converts a compiled noir contract to Aztec ABI.
+   * @param contract - A compiled noir contract.
+   * @returns Aztec ABI of the contract.
+   */
+  private convertToAztecABI(contract: NoirCompiledContract): ContractAbi {
+    return {
+      ...contract,
+      functions: contract.functions.map(noirFn => ({
+        name: noirFn.name,
+        functionType: noirFn.function_type.toLowerCase() as FunctionType,
+        parameters: noirFn.abi.parameters,
+        returnTypes: [noirFn.abi.return_type],
+        bytecode: Buffer.from(noirFn.bytecode).toString('hex'),
+        verificationKey: mockVerificationKey,
+      })),
+    };
+  }
+
+  /**
+   * Reads the dependencies of a noir crate.
+   * @param cratePath - Path to the noir crate.
+   * @returns A map of dependencies.
+   */
+  private async readDependencies(cratePath: string) {
+    const { dependencies } = toml.parse(
+      await fs.readFile(nodePath.join(cratePath, 'Nargo.toml'), { encoding: 'utf8' }),
+    );
+    return (dependencies || {}) as Record<string, Dependency>;
+  }
+
+  /**
+   * Executes the noir compiler.
+   * @returns A list of compiled noir contracts.
+   */
+  private async compileNoir(): Promise<NoirCompiledContract[]> {
+    const dependenciesMap = await this.readDependencies(this.projectPath);
+
+    noirResolver.initialiseResolver((id: string) => {
+      const idParts = id.split('/');
+
+      let path;
+      if (dependenciesMap[idParts[0]]) {
+        const [dependencyName, ...dependencySubpathParts] = idParts;
+        const dependency = dependenciesMap[idParts[0]];
+        if (!dependency.path) {
+          throw new Error(`Don't know how to resolve dependency ${dependencyName}`);
+        }
+        path = nodePath.join(this.projectPath, dependency.path, 'src', dependencySubpathParts.join('/'));
+      } else {
+        path = nodePath.join(this.projectPath, 'src', idParts.join('/'));
+      }
+
+      // The resolver does not support async resolution
+      // and holding the whole project in memory is not reasonable
+      const result = fsSync.readFileSync(path, { encoding: 'utf8' });
+      return result;
+    });
+
+    return compile({
+      contracts: true,
+      optional_dependencies_set: Object.keys(dependenciesMap), // eslint-disable-line camelcase
+    });
```

Comment:
Hold on, is the noirResolver a global object, and not an instance that we pass onto the compiler? Can we change that in noir? Global objects are always a pain.

Reply:
I'm afraid it is ): I'll check if it's possible to do in a different way on noir side, because it interacts with autogenerated code from wasm-bindgen